src/sqlfluff/core/rules/base.py

```diff
@@ -416,6 +435,59 @@ def raw_segments(self):
         )
 
 
+class CrawlBehavior:
+    """Implements how a lint rule traverses the parse tree."""
+
+    def __init__(
+        self, works_on_unparsable: bool, recurse_into: bool, needs_raw_stack: bool
+    ):
+        self.works_on_unparsable = works_on_unparsable
+        self.recurse_into = recurse_into
+        self.raw_stack: Optional[List[RawSegment]]
+        if needs_raw_stack:
+            self.raw_stack = []
+        else:
+            self.raw_stack = None  # pragma: no cover
+
+    def crawl(self, context: RuleContext) -> Iterator[RuleContext]:
+        """Yields a RuleContext for each segment the rule should process."""
+        # First, check whether we're looking at an unparsable and whether
+        # this rule will still operate on that.
+        if not self.works_on_unparsable and context.segment.is_type("unparsable"):
+            # Abort here if it doesn't. Otherwise we'll get odd results.
+            return
+
+        assert (
+            self.raw_stack is None
+            or (len(context.raw_stack) == 0 and context.raw_segment_pre is None)
+            or context.raw_stack[-1] is context.raw_segment_pre
+        )
+        # Rules should evaluate on segments FIRST, before evaluating on their
+        # children.
+        yield context
+
+        if self.recurse_into:
+            # The raw stack only keeps track of the previous *raw* segments.
+            if len(context.segment.segments) == 0:
+                context.raw_segment_pre = cast(RawSegment, context.segment)
+                # Tracking raw_stack is expensive, so it's optional per rule.
+                if self.raw_stack is not None:
+                    self.raw_stack.append(context.raw_segment_pre)
+                    context.raw_stack = tuple(self.raw_stack)
+            base_parent_stack = context.parent_stack
+            new_parent_stack = base_parent_stack + (context.segment,)
+            for idx, child in enumerate(context.segment.segments):
+                # For performance reasons, don't create a new RuleContext for
+                # each segment; just modify the existing one in place. This
+                # requires some careful bookkeeping, but it avoids creating a
+                # HUGE number of short-lived RuleContext objects
+                # (#linter loops x #rules x #segments).
+                context.segment = child
+                context.parent_stack = new_parent_stack
+                context.segment_idx = idx
+                yield from self.crawl(context)
```

Comment:
Is there a benefit to mutation. I assume its faster (cause of python), but it always gives me a bad feeling.
Is there anything preventing the crawl fn from further mutating this state and blowing things up in a completely undebuggable way?

Reply:
Performance, yes. Creating new objects is pretty expensive, and as noted in the comment above, this was probably **the** most created object in SQLFluff prior to this change (#linter loops x #rules x #segments).

Note we do create a new rule context per rule, so if a rule mutates it, it only affects that one rule, which I think helps a lot when it comes to debugging.

Reply:
We could look into having the rules access the properties through property getter functions, while the core linter modifies them by changing private fields (e.g. prefixed with an underscore) behind the scenes. Probably a later PR...